lib/puppet/configurer.rb

```diff
@@ -372,6 +372,14 @@ def run_internal(options)
       execute_postrun_command or return nil
     end
   ensure
+    if Puppet[:report_facts]
```

Comment:
A few notes:
* Can you change the name of the setting, "report_facts" to "send_facts" or "refresh_facts" or something else, it gets confusing as-is with a report being sent just a few lines down in the code at :385.
* Is it necessary to always send facts, or should we only send facts if catalog application was successful?
* If report_facts (or whatever we end up calling it) raises, then we won't send the report but we should so we need some error handling here.

Reply:
I like "report_facts" because the name indicates that the facts will be sent to the server and that the operation happens after the catalog is applied. In contrast, "refresh_facts" doesn't give any indication of what happens to the facts after they are reloaded and "send_facts" is too vague about when in the run the facts get sent.

@reidmv Any naming ideas for this one?

The failure of a single resource is unlikely to be correlated with a change in facts, so I don't think we should skip fact submission if the transaction is "failed". One case where we might skip it is if the transaction aborts due to an exception that prevents it from loading or applying any resource in the catalog. However, I don't think we expose that state in a way that is easy to detect.

Any exception raised during `report_facts` should be caught by the error trap here:

https://github.com/puppetlabs/puppet/pull/7666/files#diff-3bdc5fd6acc30b485dc30ba606c026d6R455

Reply:
If we're constraining ourselves to only two words for some reason, I agree that we kinda have to use "report" as one of them because that's currently what means "after a run" and "send to the server". Experimenting with longer names doesn't feel like it adds a whole lot of value to me.

send_updated_facts_after_agent_run
send_facts_after_run
report_updated_facts
report_facts

Pretty much boils down to what we've already got.

That said, I don't see why the method names in code need to 1:1 match the user visible setting name. If using longer or more detailed method names helps disambiguate in the code we should do that.

Reply:
I'm not a fan of `report_facts` because `report` could be a verb or a noun, where the latter is something completely different. And when combined, `report_facts` could be interpreted as we're submitting facts with the report.

How about `resubmit_facts` or `resend_facts`?

Reply:
PR updated to swap `report_facts` for `resubmit_facts`.